README.rst
Vanity
======

Get package download statistics from PyPI [1]_

.. 
    .. image:: https://travis-ci.org/aclark4life/vanity.svg
       :target: https://travis-ci.org/aclark4life/vanity
       :alt: Travis CI build status

    .. image:: https://pypip.in/v/vanity/badge.png
       :target: https://pypi.python.org/pypi/vanity/
       :alt: Latest PyPI version

    .. image:: https://pypip.in/d/vanity/badge.png
       :target: https://pypi.python.org/pypi/vanity/
       :alt: Number of PyPI downloads

    .. image:: https://coveralls.io/repos/aclark4life/vanity/badge.png?branch=travisci
       :target: https://coveralls.io/r/aclark4life/vanity?branch=travisci
       :alt: Code coverage

    .. image:: https://landscape.io/github/aclark4life/vanity/master/landscape.png
       :target: https://landscape.io/github/aclark4life/vanity/master
       :alt: Code health

Installation
------------

::

    $ pip install vanity

Usage
-----

Enter package name::

    $ vanity django
    Django-1.1.3.tar.gz    2010-12-23        4,938
    Django-1.1.4.tar.gz    2011-02-09       10,259
      Django-1.2.tar.gz    2010-05-17       24,011
    Django-1.2.1.tar.gz    2010-05-24       71,479
    Django-1.2.2.tar.gz    2010-09-09        4,388
    Django-1.2.3.tar.gz    2010-09-11       82,629
    Django-1.2.4.tar.gz    2010-12-23       66,223
    Django-1.2.5.tar.gz    2011-02-09       82,325
    Django-1.2.6.tar.gz    2011-09-10        2,559
    Django-1.2.7.tar.gz    2011-09-11       31,833
      Django-1.3.tar.gz    2011-03-23      363,202
    Django-1.3.1.tar.gz    2011-09-10      585,745
    Django-1.3.2.tar.gz    2012-07-30        7,649
    Django-1.3.3.tar.gz    2012-08-01       31,375
    Django-1.3.4.tar.gz    2013-03-05        1,974
    Django-1.3.5.tar.gz    2012-12-10       16,880
    Django-1.3.6.tar.gz    2013-02-19        2,292
    Django-1.3.7.tar.gz    2013-02-20       14,756
      Django-1.4.tar.gz    2012-03-23      437,635
    Django-1.4.1.tar.gz    2012-07-30      328,418
    Django-1.4.2.tar.gz    2012-10-17      326,088
    Django-1.4.3.tar.gz    2012-12-10      280,915
    Django-1.4.4.tar.gz    2013-02-19       12,453
    Django-1.4.5.tar.gz    2013-02-20      117,366
      Django-1.5.tar.gz    2013-02-26      124,429
    Django-1.5.1.tar.gz    2013-03-28      150,413
    ----------------------------------------------
    Django has been downloaded 3,182,234 times!

Enter package name with version::

    $ vanity pillow==2.0.0
                    Pillow-2.0.0.zip    2013-03-15       61,022
        Pillow-2.0.0.win32-py3.3.exe    2013-03-15          593
        Pillow-2.0.0.win32-py3.2.exe    2013-03-15          379
        Pillow-2.0.0.win32-py2.7.exe    2013-03-15          703
        Pillow-2.0.0.win32-py2.6.exe    2013-03-15          308
    Pillow-2.0.0.win-amd64-py3.3.exe    2013-03-15          487
    Pillow-2.0.0.win-amd64-py3.2.exe    2013-03-15          328
    Pillow-2.0.0.win-amd64-py2.7.exe    2013-03-15          500
    Pillow-2.0.0.win-amd64-py2.6.exe    2013-03-15          311
        Pillow-2.0.0-py3.3-win32.egg    2013-03-15          421
    Pillow-2.0.0-py3.3-win-amd64.egg    2013-03-15          431
        Pillow-2.0.0-py3.2-win32.egg    2013-03-15          353
    Pillow-2.0.0-py3.2-win-amd64.egg    2013-03-15          357
        Pillow-2.0.0-py2.7-win32.egg    2013-03-15        1,160
    Pillow-2.0.0-py2.7-win-amd64.egg    2013-03-15          620
        Pillow-2.0.0-py2.6-win32.egg    2013-03-15          730
    Pillow-2.0.0-py2.6-win-amd64.egg    2013-03-15          395
    -----------------------------------------------------------
    Pillow 2.0.0 has been downloaded 69,098 times!

Enter multiple package names::

    $ bin/vanity --quiet setuptools distribute
    setuptools has been downloaded 34,601,114 times!
    distribute has been downloaded 29,661,287 times!
    setuptools and distribute have been downloaded 64,262,401 times!

Enter search pattern::

    C:\>vanity pillow -p "Pillow-3.0.0.*win32.*py3.2|Pillow-3.0.0.*cp32.*win32"
        Pillow-3.0.0.win32-py3.2.exe    2015-10-01          582
    Pillow-3.0.0-cp32-none-win32.whl    2015-10-01          591
    -----------------------------------------------------------
    Pillow has been downloaded 1173 times!

.. [1] Based on https://github.com/collective/Products.PloneSoftwareCenter/commit/601558870175e35cfa4d05fb309859e580271a1f


CHANGES.rst
Changelog
=========

2.2.2 (2016-08-21)
------------------

- Fix Python 3 JSON issue
  [mattjegan]

2.2.1 (2016-04-27)
------------------

- Add message *** Note: PyPI stats are broken again; we're now waiting for warehouse. https://github.com/aclark4life/vanity/issues/22 ***

2.2.0 (2016-01-06)
------------------

- Add ``--pattern`` option
  [hugovk]

2.1.0 (2015-07-03)
------------------

- Use PyPI JSON API via ``--json``
- Replace print with logging

2.0.4 (2014-09-02)
------------------

- Query /pypi/ instead of /simple/, fixes https://github.com/aclark4life/vanity/issues/12

2.0.3 (2013-05-27)
------------------

- New life: http://mail.python.org/pipermail/distutils-sig/2013-June/021344.html

2.0.2 (2013-05-27)
------------------

- Fix typo

2.0.1 (2013-05-27)
------------------

- End of life: http://mail.python.org/pipermail/distutils-sig/2013-May/020855.html

2.0.0 (2013-05-26)
------------------

- Revert removal of ``--quiet`` option
- Support multi-package entry e.g. ``$ vanity setuptools distribute``

1.2.5 (2013-03-17)
------------------

- Switch to argparse 
- Support query by version spec e.g. ``$ vanity pillow==2.0.0``
- Remove ``--quiet`` option
- Officially add Python 3 support

1.2.4 (2013-02-19)
------------------

- Query PyPI via https
- Return usage statement when no args passed

1.2.3 (2012-08-08)
------------------

- Use optparse for option and argument parsing
  [JNRowe]
- Don't fail when the en_US locale isn't available
  [JNRowe]
- Python 3 compatibility
  [JNRowe]

1.2.2 (2012-07-31)
------------------

- Remove blessings integration which breaks on Windows

1.2.1 (2012-02-15)
------------------

- Make verbose the default
- Add blessings support to make output pretty
- install_requires requests for future refactor
- Enforce available command line options better

1.2.0 (2012-01-30)
------------------

- Add verbose option to display file name, upload date, and download count per release
- Add locale to format downloads e.g. ``700,232 times`` instead of ``700232 times``

1.1.2 (2011-10-28)
------------------

- Fix regression: Re-fix download counts
  [JNRowe]

1.1.1 (2011-10-27)
------------------

- Refactor: create ``downloads_total`` function to make external use simpler
  [kennethreitz]

1.1.0 (2011-10-25)
------------------

- Bug fixes: support for case insensitive project names and support for counting all release files (e.g. binaries in addition to sdist) and correct number of release files
  [JNRowe]

1.0 (04-13-2011)
----------------

- Initial release based on code from `Products.PloneSoftwareCenter`_ by `David Glick`_
                                                                                                                                           
.. _`Products.PloneSoftwareCenter`: https://pypi.python.org/pypi/Products.PloneSoftwareCenter
.. _`David Glick`: http://glicksoftware.com


tests.py
import unittest


class TestSuite(unittest.TestSuite):
    """
    """


pypi.py
############################################################################### 
# Copyright (C) 2011 Alex Clark                                               #
#                                                                             #
# This program is free software; you can redistribute it and/or               #
# modify it under the terms of the GNU General Public License                 #
# as published by the Free Software Foundation; either version 2              #
# of the License, or (at your option) any later version.                      #
#                                                                             #
# This program is distributed in the hope that it will be useful,             #
# but WITHOUT ANY WARRANTY; without even the implied warranty of              #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                #
# GNU General Public License for more details.                                #
#                                                                             #
# You should have received a copy of the GNU General Public License           #
# along with this program; if not, write to the Free Software                 #
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,   #
# USA.                                                                        #
###############################################################################

"""fetch download counts from PyPI"""

# Based on http://svn.plone.org/svn/collective/Products.PloneSoftwareCenter/\
# trunk/Products/PloneSoftwareCenter/pypi.py

from collections import deque
import sys
import xmlrpclib

client = xmlrpclib.ServerProxy('http://pypi.python.org/pypi')


def by_two(source):
    out = []
    for x in source:
        out.append(x)
        if len(out) == 2:
            yield out
            out = []


def package_releases(packages):
    mcall = xmlrpclib.MultiCall(client)
    called_packages = deque()
    for package in packages:
        mcall.package_releases(package, True)
        called_packages.append(package)
        if len(called_packages) == 100:
            result = mcall()
            mcall = xmlrpclib.MultiCall(client)
            for releases in result:
                yield called_packages.popleft(), releases
    result = mcall()
    for releases in result:
        yield called_packages.popleft(), releases


def release_data(packages):
    mcall = xmlrpclib.MultiCall(client)
    i = 0
    for package, releases in package_releases(packages):
        for version in releases:
            mcall.release_urls(package, version)
            mcall.release_data(package, version)
            i += 1
            if i % 50 == 49:
                result = mcall()
                mcall = xmlrpclib.MultiCall(client)
                for urls, data in by_two(result):
                    yield urls, data
    result = mcall()
    for urls, data in by_two(result):
        yield urls, data


def main():
    usage = 'Usage: vanity [my.package]'
    if len(sys.argv) >= 2 and len(sys.argv) < 3:
        if sys.argv[1] == '-h' or sys.argv[1] == '--help':
            print usage
        else:
            total = 0
            for urls, data in release_data([sys.argv[1]]):
                if not urls == []:
                    total += urls[0]['downloads']
            if total != 0:
                print 'Package `%s` has been downloaded %d times!\n' % (
                    sys.argv[1], total)
            else:
                print 'No downloads. Are you sure `%s` exists?\n' % (
                    sys.argv[1])
    else:
        print usage

if __name__ == '__main__':
    main()


vanity.py
#
# Copyright (C) 2011-2016 Alex Clark
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,
# USA.
"""
Get package download statistics from PyPI
"""

# For sorting XML-RPC results

from collections import deque, OrderedDict

# HTTPS connection for normalize function

try:
    from http.client import HTTPSConnection
except ImportError:
    from httplib import HTTPSConnection

import argparse
import locale
import logging
import json
import re
import time

# PyPI's XML-RPC methods
# https://wiki.python.org/moin/PyPIXmlRpc

try:
    import xmlrpc.client as xmlrpc
except ImportError:  # Python 2
    import xmlrpclib as xmlrpc

PYPI_HOST = 'pypi.python.org'
PYPI_URL = 'https://%s/pypi' % PYPI_HOST
PYPI_JSON = '/'.join([PYPI_URL, '%s/json'])
PYPI_XML = xmlrpc.ServerProxy(PYPI_URL)

# Print numbers with commas
# http://stackoverflow.com/a/1823101

try:
    locale.setlocale(locale.LC_ALL, 'en_US')
except locale.Error:
    pass

# Logger
# https://docs.python.org/3/howto/logging.html

logger = logging.getLogger('vanity')
logger.setLevel(logging.DEBUG)
ch = logging.StreamHandler()
ch.setLevel(logging.DEBUG)
formatter = logging.Formatter('%(message)s')
ch.setFormatter(formatter)
logger.addHandler(ch)

# PyPI JSON
# http://stackoverflow.com/a/28786650

try:
    # For Python 3.0 and later
    from urllib.request import urlopen
except ImportError:
    # Fall back to Python 2's urllib2
    from urllib2 import urlopen


def by_two(source):
    """
    """
    out = []
    for x in source:
        out.append(x)
        if len(out) == 2:
            yield out
            out = []


def count_downloads(package,
                    verbose=True,
                    version=None,
                    json=False,
                    pattern=None):
    """
    """
    count = 0
    items = []
    for urls, data in get_release_info([package], json=json):
        for url in urls:
            filename = url['filename']
            if pattern and not re.search(pattern, filename):
                continue
            downloads = url['downloads']
            downloads = locale.format("%d", downloads, grouping=True)
            if not json:
                upload_time = url['upload_time'].timetuple()
                upload_time = time.strftime('%Y-%m-%d', upload_time)
            else:
                # Convert 2011-04-14T02:16:55 to 2011-04-14
                upload_time = url['upload_time'].split('T')[0]
            if version == data['version'] or not version:
                item = '%s    %s    %9s' % (filename, upload_time, downloads)
                items.append(item)
                count += url['downloads']
    if verbose and items != []:
        items.reverse()
        # http://stackoverflow.com/questions/873327/\
        # pythons-most-efficient-way-to-choose-longest-string-in-list
        longest = len(max(items, key=len))
        for item in items:
            logger.debug(item.rjust(longest))
        logger.debug('-' * longest)
    return count


# http://stackoverflow.com/a/28786650
def get_jsonparsed_data(url):
    """Receive the content of ``url``, parse it as JSON and return the
       object.
    """
    response = urlopen(url)
    response = json.loads(response.read().decode('utf-8'))

    sorted_releases = OrderedDict()
    for release in sorted(response['releases'].keys())[::-1]:
        sorted_releases[release] = response['releases'][release]
    response['releases'] = sorted_releases
    return response


def normalize(name):
    """
    """
    http = HTTPSConnection(PYPI_HOST)
    http.request('HEAD', '/pypi/%s/' % name)
    r = http.getresponse()
    if r.status not in (200, 301):
        raise ValueError(r.reason)
    return r.getheader('location', name).split('/')[-1]


def get_releases(packages):
    """
    """
    mcall = xmlrpc.MultiCall(PYPI_XML)
    called_packages = deque()
    for package in packages:
        mcall.package_releases(package, True)
        called_packages.append(package)
        if len(called_packages) == 100:
            result = mcall()
            mcall = xmlrpc.MultiCall(PYPI_XML)
            for releases in result:
                yield called_packages.popleft(), releases
    result = mcall()
    for releases in result:
        yield called_packages.popleft(), releases


def get_release_info(packages, json=False):
    """
    """
    if json:
        for package in packages:
            data = get_jsonparsed_data(PYPI_JSON % package)
            for release in data['releases']:
                urls = data['releases'][release]
                yield urls, data['info']
        return

    mcall = xmlrpc.MultiCall(PYPI_XML)

    i = 0
    for package, releases in get_releases(packages):
        for version in releases:
            mcall.release_urls(package, version)
            mcall.release_data(package, version)
            i += 1
            if i % 50 == 49:
                result = mcall()
                mcall = xmlrpc.MultiCall(PYPI_XML)
                for urls, data in by_two(result):
                    yield urls, data

    result = mcall()
    for urls, data in by_two(result):
        yield urls, data


def vanity():
    """
    """
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('package', help='pypi package name', nargs='+')
    parser.add_argument('-q',
                        '--quiet',
                        help='only show total downloads',
                        action='store_true')
    parser.add_argument('-j',
                        '--json',
                        help='use pypi json api instead of xmlrpc',
                        action='store_true')
    parser.add_argument('-p',
                        '--pattern',
                        help='only show files matching a regex pattern')
    args = parser.parse_args()
    packages = args.package
    verbose = not (args.quiet)
    version = None
    grand_total = 0
    package_list = []
    json = args.json
    for package in packages:
        if package.find('==') >= 0:
            package, version = package.split('==')
        try:
            package = normalize(package)
        except ValueError:
            parser.error('No such module or package %r' % package)

        # Count downloads
        total = count_downloads(package,
                                json=json,
                                version=version,
                                verbose=verbose,
                                pattern=args.pattern)
        if total != 0:
            if version:
                logger.debug('%s %s has been downloaded %s times!' %
                             (package, version, locale.format("%d",
                                                              total,
                                                              grouping=True)))
            else:
                logger.debug('%s has been downloaded %s times!' %
                             (package, locale.format("%d",
                                                     total,
                                                     grouping=True)))
        else:
            if version:
                logger.debug('No downloads for %s %s.' % (package, version))
            else:
                logger.debug('No downloads for %s.' % package)
        grand_total += total
        package_list.append(package)
    if len(package_list) > 1:
        package_string = (
            ', '.join(package_list[:-1]) + " and " + package_list[-1])
        logger.debug("%s have been downloaded %s times!" %
                     (package_string, locale.format("%d",
                                                    grand_total,
                                                    grouping=True)))

    logger.debug("\n\n\t *** Note: PyPI stats are broken again; we're now waiting for warehouse. https://github.com/aclark4life/vanity/issues/22 ***\n\n")

if __name__ == '__main__':
    vanity()
